smart-contracts/contracts/Unlock.sol

```diff
@@ -229,8 +244,39 @@ contract Unlock is
         valueInETH = _value;
       }
 
-      grossNetworkProduct += valueInETH;
+      grossNetworkProduct = grossNetworkProduct.add(valueInETH);
+      // If GNP does not overflow, the lock totalSales should be safe
       locks[msg.sender].totalSales += valueInETH;
+
+      // Mint UDT
+      if(_referrer != address(0))
+      {
+        IUniswapOracle udtOracle = uniswapOracles[address(udt)];
+        if(address(udtOracle) != address(0))
+        {
+          // Get the value of 1 UDT (w/ 18 decimals) in ETH
+          uint udtPrice = udtOracle.updateAndConsult(address(udt), 10 ** 18, weth);
+          uint gasCost = estimatedGasForPurchase * tx.gasprice;
+          uint tokensToMint = gasCost / udtPrice;
+          // If the delta is < 0.01% then no tokens will be minted
+          uint percentGNPChangeInBP = BASIS_POINTS_DEN.mul(valueInETH) / grossNetworkProduct;
+          uint maxTokens = percentGNPChangeInBP.mul(udt.totalSupply()) / BASIS_POINTS_DEN;
```

Comment:
We should divide `percentGNPChangeInBP` by 2, because the supply can only increase at half the speed of the GNP.